source/common/config/delta_subscription_state.cc

```diff
@@ -14,37 +14,46 @@ namespace Config {
 DeltaSubscriptionState::DeltaSubscriptionState(std::string type_url,
                                                UntypedConfigUpdateCallbacks& watch_map,
                                                const LocalInfo::LocalInfo& local_info,
-                                               Event::Dispatcher& dispatcher, const bool wildcard)
+                                               Event::Dispatcher& dispatcher)
     // TODO(snowp): Hard coding VHDS here is temporary until we can move it away from relying on
     // empty resources as updates.
     : supports_heartbeats_(type_url != "envoy.config.route.v3.VirtualHost"),
       ttl_(
           [this](const auto& expired) {
             Protobuf::RepeatedPtrField<std::string> removed_resources;
             for (const auto& resource : expired) {
-              setResourceWaitingForServer(resource);
-              removed_resources.Add(std::string(resource));
+              if (auto maybe_resource = getRequestedResourceState(resource);
+                  maybe_resource.has_value()) {
+                maybe_resource->setAsWaitingForServer();
+                removed_resources.Add(std::string(resource));
+              } else if (auto erased_count = wildcard_resource_state_.erase(resource);
+                         erased_count > 0) {
+                removed_resources.Add(std::string(resource));
+              }
             }
 
             watch_map_.onConfigUpdate({}, removed_resources, "");
           },
           dispatcher, dispatcher.timeSource()),
-      type_url_(std::move(type_url)), wildcard_(wildcard), watch_map_(watch_map),
-      local_info_(local_info), dispatcher_(dispatcher) {}
+      type_url_(std::move(type_url)), watch_map_(watch_map), local_info_(local_info),
+      dispatcher_(dispatcher) {}
 
 void DeltaSubscriptionState::updateSubscriptionInterest(
     const absl::flat_hash_set<std::string>& cur_added,
     const absl::flat_hash_set<std::string>& cur_removed) {
   for (const auto& a : cur_added) {
-    setResourceWaitingForServer(a);
+    // This adds a resource state that is waiting for the server for more information. This also may
+    // be a wildcard resource, which is fine too.
+    requested_resource_state_.insert_or_assign(a, ResourceState::waitingForServer());
```

Comment:
Is there a way for a resource to transition back from subscribed to wildcard?

Reply:
Hm, not directly I think. Should the client be smart and move the resource from subscribed to wildcard or should the client rely on the server detecting that the client unsubscribed from a resource that is also a part of the wildcard subscription and thus resending the resource? I wonder if this scenario should be clarified in the xds-protocol document.

Reply:
Transitioning from subscribed to wildcard is also tricky. Consider the following scenarios:

1. Client subscribes to wildcard.
2. Server sends resource `r` as a part of wildcard subscription.
3. Client subscribes to `r`. At this point the client may know that `r` also came from wildcard subscription.
4. Client unsubscribes from `r`, but the client can still keep `r` as a resource coming from wildcard subscription.

Another one:
1. Client subscribes to `r` and wildcard.
2. Server sends resource `r` as a part of both wildcard subscription and explicit subscription to `r`.
3. Client unsubscribes from `r`. At this point the client does not know that `r` is also a part of the wildcard subscription.

So I'd say that the client should rather rely on the server. Basically the server should notice that the client is subscribed to the resource that is also a part of the wildcard resource set, so when the client unsubscribes from such a resource, the server should resend the resource, so the client could get it as a part of the wildcard subscription set.

Other option would be to extend the discovery response with a boolean stating whether the resource is a part of the wildcard resource set too. This could be used by the client to transition the resource from subscribed to wildcard when unsubscribing from the resource.

Reply:
I don't think that the transport protocol actually provides enough information for the client to know that the server is sending the resource as part of a wildcard subscription.  Your second scenario is a good example of this.  Here's another one:

1. Client sends a request subscribing to wildcard.
2. Client sends a request subscribing to both wildcard and `r`.
3. Client receives the server's response to its initial request subscribing to wildcard, which includes `r`.

Because the client did not receive the response to the first request until after it sent the second request, it has no way to know whether the response it's seeing is a result of the first request or the second, so it cannot assume that `r` is present because of the wildcard request.

I think the right answer here is to just say that when a client is sending a wildcard subscription, it should never delete resources from its cache when it unsubscribes from a non-wildcard resource on the same stream; instead, it should rely on the resource being removed from the server's next response to do that.  (Note that in LDS and CDS, the only two resource types where wildcards are used, the server is required to send all resources that the client is subscribed to in every response, so if a resource is present in one response and absent in the next response, that's a clear indication that the client no longer needs the resource.)